### DISCUSSION TITLE
Pattern-based commands and specific actions for empty commands

### DISCUSSION THREAD

**Discussion by blairfrandeen** [python-cmd2/cmd2#1246] [Q&A] [answered]
I'm writing an application that I use to track my working hours on various projects, and I'm trying to decide if using `cmd2` makes sense for my specific use case. In exploring the documentation and examples, I've found two features of my current application that `cmd2` doesn't seem to support: Pattern-based commands, and calling a specific function for an empty command.

The principle behind my time tracking application is minimum amount of keystrokes to enter data. Thus if I want to log an activity, any command that starts with a float logs that many hours. Optional arguments can follow that and are parsed accordingly. For example, a command such as `1.25 2 t example command` is parsed as "`1.25` hours on work mode `2` (deep work) on project code `t` (time tracker app) with comment `example command.`" 

Having optional arguments in `cmd2` is easy enough, but can I run these without any command to begin with? Having to type `add_entry` prior to any arguments is not an improvement, and `cmd2` doesn't accept `def do_():` as a function declaration or `""` as an alias.

Another feature of this application is that it reads my Outlook calendar, and prompts me to make time entries based on that. For example, if I have a 1 hour team meeting on my outlook calendar, the application will prompt me as to whether that was correct. If it was, I simply press `<enter>` and the details from the Outlook event are entered in my time log. If it wasn't, I can use arguments as described above to modify the entry before its recorded.

Is there a way to use an empty command to do something specific, based on the state of the program?

I've implemented all of the above with my own console module in my project called [titr](https://github.com/blairfrandeen/titr). It works, but the console is rough around the edges, and doesn't have some of the nicer features of `cmd2` such as command history. The console is fundamentally different than `cmd2` in that it uses function decorators to collect console commands, so switching to `cmd2` would mean a major refactoring effort.

Looking forward to hearing any thoughts and feedback.

**Comment by anselor** [comment, marked as answer]
This behavior is very different from how most of cmd behaves. There are 2 avenues you could explore, though.

In the `Cmd` class, all commands eventually hit the `onecmd()` function where it attempts to locate the `do_` function that corresponds to a command. If it can't find a command, it calls the `default()` function. You could try overriding `default()` and, failing that, try overriding `onecmd()`. If you have custom state tracking, you can probably use as the basis for determining which of your functions you want to call.

Completion starts by calling the `Cmd.complete()` function which then looks up the command function and determines the type of completion depending on the specifics of the command function. 

As for the note about decorators vs `do_` functions - that's a holdover from cmd2 being built on top of python's Cmd which depends on these do_ functions. It may be possible for us to support a decorator based command registration in the future.

**Comment by blairfrandeen** [comment]
Thanks, this pointed me in the right direction. I was able to get the pattern-based behavior by modifying `onecmd`, and the empty-command behavior by modifying `_complete_statement`. The latter is a bit hacky in that I have to call `self.statement_parser.parse()` with my preferred function as the argument, but it seems to work.
